### PR TITLE
Added validation of the host name in mqtt.client.connect()

### PR DESCRIPTION
**Pull request by dreamiurg**
Signed-off-by: Dmitry Guyvoronsky <dmitry.guyvoronsky@gmail.com>

### CODE REVIEW

**Comment by Nzbuu** [comment]
Is it necessary to hard-code this logic here? What about IPv6 addresses?

Also, does it add anything over just waiting for `socket.create_connection` to fail instead?

**Comment by dreamiurg** [comment]
@Nzbuu, generally I would I prefer EAFP over LBYL (look-before-you-leap). However, in this case i thought that host name validation would give user more detailed error message. We've seen users passing incorrect hostname to the `connect()`, e.g `mqtt.client.connect("https://mqtt-endpoint.example.com", ...)` that lead to  `socket.getaddrinfo()` failing with pretty generic
`socket.gaierror: [Errno -2] Name or service not known` and caused some confusion.

What do you think about validating host name before trying to connect to it?

**Comment by Nzbuu** [comment]
@dreamiurg, if your objective is just to throw a better exception, then my preference would be to do it in the `catch` block associated with opening the socket. As far as I am concerned, it's unnecessary to pre-check the host name, when `socket.create_connection` should be responsible for that. Especially considering that the checks that you have added are almost certainly not the same, for example IPv6 addresses are forbidden by this code and the `socket` package apply different constraints on the host name.

File: src/paho/mqtt/client.py [hunk 1]
```diff
@@ -706,8 +706,19 @@ def connect_async(self, host, port=1883, keepalive=60, bind_address=""):
         broker. If no other messages are being exchanged, this controls the
         rate at which the client will send ping messages to the broker.
         """
-        if host is None or len(host) == 0:
-            raise ValueError('Invalid host.')
+        import re
```
Comment (Nzbuu) [inline src/paho/mqtt/client.py]: Would it be better to add this in the global imports? A user will always call this method at least once.

File: src/paho/mqtt/client.py [hunk 2]
```diff
@@ -706,8 +706,19 @@ def connect_async(self, host, port=1883, keepalive=60, bind_address=""):
...
+        if host is None or len(host) == 0 or len(host) > 255:
+            raise ValueError('Invalid host "{0}": length should be between 1 and 255 characters.'.format(host))
+        if host[-1] == ".":
+            host = host[:-1] # strip exactly one dot from the right, if present
+
+        RE_VALID_IP_ADDRESS = re.compile("^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$", re.IGNORECASE)
+        RE_VALID_HOSTNAME = re.compile("^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$", re.IGNORECASE)
+
+        if not RE_VALID_IP_ADDRESS.match(host) and not RE_VALID_HOSTNAME.match(host):
+            raise ValueError('Invalid host "{0}": does not appear to be neither IP address nor hostname.'.format(host))
```
Comment (Nzbuu) [inline src/paho/mqtt/client.py]: This is very hard to test. I think this would be better extracted as a method or function.

**Comment by dreamiurg** [comment]
After thinking about this for some more time, I think that pre-validation is not necessary. I'll explore other ways to give users more detailed error information instead.

**Pull request closed without merging**